pkg/controller/replication_controller_utils.go

```diff
@@ -0,0 +1,270 @@
+	if e, ok := obj.(*PodExpectations); ok {
+		return e.key, nil
+	}
+	return "", fmt.Errorf("Could not find key for obj %#v", obj)
+}
+
+// RCExpectations is a ttl cache mapping rcs to what they expect to see before being woken up for a sync.
+type RCExpectations struct {
+	cache.Store
+}
+
+// GetExpectations returns the PodExpectations of the given rc.
+func (r *RCExpectations) GetExpectations(rc *api.ReplicationController) (*PodExpectations, bool, error) {
+	rcKey, err := rcKeyFunc(rc)
+	if err != nil {
+		return nil, false, err
+	}
+	if podExp, exists, err := r.GetByKey(rcKey); err == nil && exists {
+		return podExp.(*PodExpectations), true, nil
+	} else {
+		return nil, false, err
+	}
+}
+
+// CheckExpectations returns true if the replication manager has observed the required adds/dels
+// for the given rc. Add/del counts are established by the rc at sync time, and updated as pods
+// are observed by the replication manager's podController.
+func (r *RCExpectations) CheckExpectations(rc *api.ReplicationController) bool {
+	if podExp, exists, err := r.GetExpectations(rc); exists {
+		if podExp.Fulfilled() {
+			return true
+		} else {
+			glog.V(4).Infof("Controller %v still waiting on expectations %#v", podExp)
+			return false
+		}
+	} else if err != nil {
+		glog.V(2).Infof("Error encountered while checking expectations %#v, forcing sync", err)
+	}
+	// Trigger a sync if we either encountered and error (which shouldn't happen since we're
+	// getting from local store) or this rc hasn't established expectations.
+	return true
+}
+
+// SetExpectations registers new expectations for the given rc. Forgets existing expectations.
+func (r *RCExpectations) SetExpectations(rc *api.ReplicationController, add, del int) error {
+	rcKey, err := rcKeyFunc(rc)
+	if err != nil {
+		return err
+	}
+	return r.Add(&PodExpectations{add: int64(add), del: int64(del), key: rcKey})
+}
+
+// Decrements the expectation counts of the given rc.
+func (r *RCExpectations) lowerExpectations(rc *api.ReplicationController, add, del int) {
+	if podExp, exists, err := r.GetExpectations(rc); err == nil && exists {
+		podExp.Seen(int64(add), int64(del))
+	}
+}
+
+// CreationObserved atomically decrements the `add` expecation count of the given replication controller.
+func (r *RCExpectations) CreationObserved(rc *api.ReplicationController) {
+	r.lowerExpectations(rc, 1, 0)
+}
+
+// DeletionObserved atomically decrements the `del` expectation count of the given replication controller.
+func (r *RCExpectations) DeletionObserved(rc *api.ReplicationController) {
+	r.lowerExpectations(rc, 0, 1)
+}
+
+// Expectations are either fulfilled, or expire naturally.
+type Expectations interface {
+	Fulfilled() bool
+}
+
+// PodExpectations track pod creates/deletes.
+type PodExpectations struct {
+	add int64
+	del int64
+	key string
+}
+
+// Seen decrements the add and del counters.
+func (e *PodExpectations) Seen(add, del int64) {
```

Comment:
I wanted to allow the use of podExpectations through anything that isn't an RCExpectations but can move it into lowerExpectations if you prefer. Will make it private.